File: src/QuickSort.ts
const qs = (array: number[], low: number, high: number): void => {
  if (low >= high) {
    console.log(array);
    return;
  }

  const pivotIndex = partition(array, low, high);

  qs(array, low, pivotIndex - 1);
  qs(array, pivotIndex + 1, high);
};

const partition = (array: number[], low: number, high: number): number => {
  const pivot = array[high];
  let index = low - 1;

  for (let i = low; i < high; i++) {
    if (array[i] <= pivot) {
      index++;
      const temp = array[i];
      array[i] = array[index];
      array[index] = temp;
    }
  }
  index++;
  array[high] = array[index];
  array[index] = pivot;

  return index;
};

export const quick_sort = (array: number[]): void => {
  qs(array, 0, array.length - 1);
};

quick_sort([9, 3, 7, 4, 69, 420, 42]);

File: rust/QuickSort.rs
fn quick_sort(array: &mut [u8], low: usize, high: usize) {
    if low >= high {
        return;
    }

    let pivot: u8 = array[high];
    let mut i = low;

    for j in low..high {
        if array[j] <= pivot {
            array.swap(i, j);
            i += 1;
        }
    }

    array.swap(i, high);

    quick_sort(array, low, i - 1);
    quick_sort(array, low + 1, high);
}

fn main() {
    let mut arr: [u8; 9] = [77, 6, 35, 2, 3, 32, 11, 25, 9];
    let high = arr.len() - 1;

    quick_sort(&mut arr, 0, high);

    println!("Ordered array : {:?}", arr);
}
